src/main/scala/com/databricks/spark/sql/perf/mllib/BenchmarkAlgorithm.scala

```diff
@@ -0,0 +1,94 @@
+package com.databricks.spark.sql.perf.mllib
+
+import com.typesafe.scalalogging.slf4j.Logging
+
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.evaluation.{MulticlassClassificationEvaluator, Evaluator}
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
+
+/**
+ * The description of a benchmark for an ML algorithm. It follows a simple, standard proceduce:
+ *  - generate some test and training data
+ *  - generate a model against the training data
+ *  - score the model against the training data
+ *  - score the model against the test data
+ *
+ * You should not assume that your implementation can carry state around. If some state is needed,
+ * consider adding it to the context.
+ *
+ * It is assumed that the implementation is going to be an object.
+ */
+trait BenchmarkAlgorithm extends Logging {
+
+  def trainingDataSet(ctx: MLBenchContext): DataFrame
+
+  def testDataSet(ctx: MLBenchContext): DataFrame
+
+  @throws[Exception]("if training fails")
+  def train(
+      ctx: MLBenchContext,
+      trainingSet: DataFrame): Transformer
+
+  @throws[Exception]("if scoring fails")
+  def score(
+      ctx: MLBenchContext,
+      testSet: DataFrame,
+      model: Transformer): Double = -1.0 // Not putting NaN because it is not valid JSON.
+}
+
+/**
+ * Uses an evaluator to perform the scoring.
+ */
+trait ScoringWithEvaluator {
+  self: BenchmarkAlgorithm =>
+
+  protected def evaluator(ctx: MLBenchContext): Evaluator = new MulticlassClassificationEvaluator()
```

Comment:
Moving the default implementation to logistic regression. This is only one line of code that we can factorize later anyway.